File: src/codes/excessive-branching/ts/rfc_command.ts
// コマンドインターフェース
interface DiscountCommand {
  execute(basePrice: number): number;
}

// 各コマンド（ConcreteCommand）
class StudentDiscountCommand implements DiscountCommand {
  execute(basePrice: number): number {
    return basePrice * 0.8;
  }
}

class MemberDiscountCommand implements DiscountCommand {
  execute(basePrice: number): number {
    return basePrice * 0.9;
  }
}

class VIPDiscountCommand implements DiscountCommand {
  execute(basePrice: number): number {
    return basePrice * 0.7;
  }
}

class NoDiscountCommand implements DiscountCommand {
  execute(basePrice: number): number {
    return basePrice;
  }
}

// Invoker
class DiscountExecutor {
  constructor(private command: DiscountCommand) {}

  setCommand(command: DiscountCommand) {
    this.command = command;
  }

  run(basePrice: number): number {
    return this.command.execute(basePrice);
  }
}

// 利用例
const userType = "member";
let command: DiscountCommand;

switch (userType) {
  case "student":
    command = new StudentDiscountCommand();
    break;
  case "member":
    command = new MemberDiscountCommand();
    break;
  case "vip":
    command = new VIPDiscountCommand();
    break;
  default:
    command = new NoDiscountCommand();
}

const executor = new DiscountExecutor(command);
console.log(executor.run(1000)); // 900

File: src/codes/excessive-branching/py/rfc_command.py
from abc import ABC, abstractmethod

class DiscountCommand(ABC):
    @abstractmethod
    def execute(self, base_price: float) -> float:
        pass

class StudentDiscountCommand(DiscountCommand):
    def execute(self, base_price: float) -> float:
        return base_price * 0.8

class MemberDiscountCommand(DiscountCommand):
    def execute(self, base_price: float) -> float:
        return base_price * 0.9

class VIPDiscountCommand(DiscountCommand):
    def execute(self, base_price: float) -> float:
        return base_price * 0.7

class NoDiscountCommand(DiscountCommand):
    def execute(self, base_price: float) -> float:
        return base_price

class DiscountExecutor:
    def __init__(self, command: DiscountCommand):
        self.command = command

    def set_command(self, command: DiscountCommand):
        self.command = command

    def run(self, base_price: float) -> float:
        return self.command.execute(base_price)

# 利用例
user_type = "member"

if user_type == "student":
    command = StudentDiscountCommand()
elif user_type == "member":
    command = MemberDiscountCommand()
elif user_type == "vip":
    command = VIPDiscountCommand()
else:
    command = NoDiscountCommand()

executor = DiscountExecutor(command)
print(executor.run(1000))  # 900
